### ISSUE TITLE
WebGUI doesn't start on 24.1.10 and 24.7

### ISSUE THREAD

**Issue by CatFoxVoyager** [opnsense/core#7649]
Issue Summary

Problem: The WebGUI in OPNsense versions 24.1.10 and later (including 24.7) fails to work properly. Reverting to version 24.1.8 resolves the issue.

Steps to Reproduce

    Upgrade OPNsense to version 24.1.10 or later.
    Attempt to access the WebGUI.
    Observe that the WebGUI does not load
    

OPNsense 24.1.8-amd64
FreeBSD 13.2-RELEASE-p11
OpenSSL 3.0.14

    
   Log :
   
   `2024-07-25 15:32:48: (/usr/obj/usr/ports/www/lighttpd/work/lighttpd-1.4.76/src/mod_openssl.c.1453) SSL: inactive/expired X509 certificate '/var/etc/cert.pem'
2024-07-25 15:32:48: (/usr/obj/usr/ports/www/lighttpd/work/lighttpd-1.4.76/src/network.c.604) bind() [fe80::200:b1ff:fe4d:0]:443: Can't assign requested address
...
2024-07-25T15:32:48-04:00 opnsense 296 - [meta sequenceId="305"] <118>Starting web GUI...failed.
`

After doing `configctl webgui restart renew`

`event @ 1721936304.36 exec: system event config_changed response: OK  
<171>1 2024-07-25T15:38:44-04:00 opn.exemple.com config 454 - [meta sequenceId="4"] /usr/local/etc/rc.restart_webgui: The command '/usr/local/bin/flock -ne /var/run/lighty-webConfigurator.pid /usr/local/sbin/lighttpd -f /var/etc/lighty-webConfigurator.conf' returned exit code '255', the output was '2024-07-25 15:38:44: (/usr/obj/usr/ports/www/lighttpd/work/lighttpd-1.4.76/src/network.c.604) bind() [fe80::200:b1ff:fe4d:0]:443: Can't assign requested address'`

configt of lighttpd : [lighty-webConfigurator.txt](https://github.com/user-attachments/files/16383071/lighty-webConfigurator.txt)

I've Suricata, got some error after upgrading to 24.1.10 and 24.7

'<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com  kernel - - [meta sequenceId="419"] <6>vlan100: permanently promiscuous mode enabled
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="420"] 877.066913 [1171] generic_netmap_attach     Emulated adapter for vlan100 created (prev was NULL)
<13>1 2024-07-25T15:14:37-04:00opn.exemple.com kernel - - [meta sequenceId="421"] 877.066980 [ 319] generic_netmap_register   Emulated adapter for vlan100 activated
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="422"] 877.074053 [1171] generic_netmap_attach     Emulated adapter for pppoe1 created (prev was NULL)
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="423"] 877.074067 [1076] generic_netmap_dtor       Emulated netmap adapter for pppoe1 destroyed
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="424"] 877.074084 [1171] generic_netmap_attach     Emulated adapter for pppoe1 created (prev was NULL)
<13>1 2024-07-25T15:14:37-04:00 opn.jedduff.com kernel - - [meta sequenceId="425"] 877.074092 [1076] generic_netmap_dtor       Emulated netmap adapter for pppoe1 destroyed
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="426"] 877.074105 [1171] generic_netmap_attach     Emulated adapter for pppoe1 created (prev was NULL)
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="427"] 877.074116 [1076] generic_netmap_dtor       Emulated netmap adapter for pppoe1 destroyed
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="428"] <6>pppoe1: permanently promiscuous mode enabled
<13>1 2024-07-25T15:14:37-04:00opn.exemple.comkernel - - [meta sequenceId="429"] 877.076078 [1171] generic_netmap_attach     Emulated adapter for pppoe1 created (prev was NULL)
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="430"] 877.076164 [ 319] generic_netmap_register   Emulated adapter for pppoe1 activated
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="431"] <118>>>> Invoking start script 'cron'
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="432"] <118>Starting Cron: OK
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="433"] <118>
<13>1 2024-07-25T15:14:37-04:00 opn.exemple.com kernel - - [meta sequenceId="434"] <118>'

I did a opnsense-revert -r 21.1.8 and I don't need to reboot and if i'm doing configctl webgui restart, i can access the gui right away.

**Comment by fichtner** [comment]
Pretty sure you set explicit listen interfaces in the GUI and didn’t heed the advice in the GUI and the documentation. This has been discussed countless times. 😉

**Comment by CatFoxVoyager** [comment]
@fichtner You are right, I forgot about that. I've set interface listening to all and everything is ok now. Weird behavior between version tho.

**Comment by fichtner** [comment]
Well, there is a warning dialog when you set anything other than „all (recommended)“ and there is a dedicated section in the documentation https://docs.opnsense.org/manual/settingsmenu.html „listening interfaces“.

Note this has nothing to do with updates or upgrades.

**Comment by CatFoxVoyager** [comment]
Nothing to do with upgrade/update?

I deploy a fresh 24.7 and webgui crash after importing my backup. 

I revert to 24.1 (fresh install), no problem until 24.1.10.

So yeah, some code of a library got changed after 24.1.10, so obviously, there is an issue somehow with the upgrade/update.

**Comment by fichtner** [comment]
Frankly, I fail to see the problem here? You set it up wrong. People have asked for this feature but it’s not reliable. It has been documented as such.

**Comment by AdSchellevis** [comment]
As usual, everybody is free to ignore all warnings, even though they have been there for many years (https://github.com/opnsense/core/commit/795dd8bd72860ec39298347421e62001aca6fa23), this includes documentation on how to use features like these https://docs.opnsense.org/manual/settingsmenu.html#listen-interfaces.

**Comment by fichtner** [comment]
Sure, feel free to propose a better wording.

**Comment by tionu** [comment]
As a fix I'm thinking of adding a "dad_delay" as in [rc.newwanipv6#L71](https://github.com/opnsense/core/blob/9539cf27e4d273f0a4f435e7b11e92bed8cb503d/src/etc/rc.newwanipv6#L71) to [webgui.inc#L64](https://github.com/opnsense/core/blob/9539cf27e4d273f0a4f435e7b11e92bed8cb503d/src/etc/inc/plugins.inc.d/webgui.inc#L64). This should only affect users who have explicitly set listening interfaces at **System: Settings: Administration** for the WebGUI.
Any feedback, if such a PR would be welcome?

**Comment by fichtner** [comment]
@tionu Why should "dad_delay" fix it? tentative addresses are not eligible for binding so they are ignored by interfaces_addresses(). I suspect a different issue and we're trying to fix the wrong end.

**Comment by fichtner** [comment]
FWIW, I improved the locking in webgui.inc via 4182f19 which may help with certain race conditions, but likely not with this binding one.

    # opnsense-patch 4182f1993

This is actually one of the cases that will never reliably work since we're reading addresses that lighttpd considers static so when they are not it simply fails which is correct. It doesn't really matter if that's the system fiddling with link-local addresses or user configuration error or something else. Forcing a list of dynamic addresses is dangerous. The best option to force is using a loopback device with a fixed IPv4 address only unless you really need an IPv6 in which case `net.inet6.ip6.dad_count=0` is your best tunable friend.


Cheers,
Franco

**Comment by sxc731** [comment]
> Coming from other FW products, selecting an interface makes sense, and I know the correct interface I want to listen on.

This also happens to be [Defense in Depth](https://en.wikipedia.org/wiki/Defense_in_depth_(computing)) 101...  Something a firewall product should support, if not outright encourage IMHO.

**Comment by fichtner** [comment]
Why oh why is everyone arguing around the fact that this is the case in the default install? You can only access LAN. 🤷‍♂️

**Comment by tionu** [comment]
@fichtner I was assuming a DAD race condition because setting the tunable „net.inet6.ip6.dad_count“ to „0“ solves the bind issue.

I've tried to find in [interfaces_addresses()](https://github.com/opnsense/core/blob/62a09bfa0a189988b66317db93539dbf49426e5d/src/etc/inc/interfaces.inc#L4136) where tentative addresses are ignored. Seems as if the status is just returned, but the address is not omitted from the result.

The following change to [webgui.inc#L85](https://github.com/opnsense/core/blob/62a09bfa0a189988b66317db93539dbf49426e5d/src/etc/inc/plugins.inc.d/webgui.inc#L85) filters tentative addresses and fixes the issue with lighttpd trying to bind to them.

```php
foreach (interfaces_addresses($interfaces) as $tmpaddr => $info) {
        if ($info['tentative']) {
            service_log('Skipping tentative address: ' . $tmpaddr, $verbose);
            continue;
        }

        if (!$info['bind']) {
            continue;
        }

        $listeners[] = $tmpaddr;
    }
```

Maybe this would be a more suitable approach than adding a "dad_delay"?

**Comment by fichtner** [comment]
Ok I was thinking of:

https://github.com/opnsense/core/blob/62a09bfa0a189988b66317db93539dbf49426e5d/src/etc/inc/interfaces.inc#L4250

But that's not interfaces_addresses(). Don't mind ignoring tentative and it doesn't need a log message IMO. You can add it to the 'bind' conditional. Simple one line change.

**Comment by fichtner** [comment]
And for the record I forgot to note this bind failure never happened on my IPv6 install so it's either hardware and/or configuration related which makes this even harder find out what the actual issue is. The address shouldn't be tentative at that point anymore.

**Comment by tionu** [comment]
How about:

```php
    foreach (interfaces_addresses($interfaces) as $tmpaddr => $info) {
        if (!$info['bind'] || ($info['family'] == 'inet6' && $info['tentative'])) {
            continue;
        }

        $listeners[] = $tmpaddr;
    }
```

**Comment by tionu** [comment]
Submitted a pull request: https://github.com/opnsense/core/pull/7731

**Comment by fichtner** [comment]
While the commit will prevent breakage all of this feels fishy. I have a suspicion around `ifdisabled` flag toggling but looking at 24.1.8 vs. 24.1.9 I don't see the expected code has changed in that regard.

**Comment by fichtner** [comment]
Hmm, the only thing I can think of is a timing difference between running the two sysctls vs. not running them. We do want the boot to be faster, but then we also don't want to add arbitrary delays to fix the kernel auto-link-local feature to go out of sync because it's not ready.

One thing that could be considered is briefly allowing `no_dad` to work its magic and then turning it back on immediately. That way the link-local wouldn't appear as tentative, but it's a slippery slope. We've done something similar for static IPv6 addressing setting `no_dad` because it gets in the way in other situations as well.

**Comment by fichtner** [comment]
Since all of this remains sketchy I'll ship the mildly relevant locking fix 74d2a5a in 24.7.2 and that should be ok.
